firedrake/function.py

```diff
@@ -354,13 +355,22 @@ def _interpolate(self, fs, dat, expression, subset):
             raise RuntimeError('Rank mismatch: Expression rank %d, FunctionSpace rank %d'
                                % (expression.rank(), len(fs.ufl_element().value_shape())))
 
-        if expression.value_shape() != fs.ufl_element().value_shape():
+        if expression.ufl_shape != fs.ufl_element().value_shape():
             raise RuntimeError('Shape mismatch: Expression shape %r, FunctionSpace shape %r'
-                               % (expression.value_shape(), fs.ufl_element().value_shape()))
+                               % (expression.ufl_shape, fs.ufl_element().value_shape()))
 
         coords = fs.mesh().coordinates
 
-        if hasattr(expression, "eval"):
+        if not isinstance(expression, expression_t.Expression):
+            kernel, oriented, coefficients = self._interpolate_ufl_expression(expression, to_pts, to_element, fs)
+            args = [kernel, subset or self.cell_set,
+                    dat(op2.WRITE, fs.cell_node_map()[op2.i[0]])]
```

Comment:
OK, I see what's going on.  The first index to the output tensor is not a loop over basis function values in the same way, and therefore doesn't need flattening (the unrolling happened in the kernel).